New resource `aws_vpc_endpoint_security_group_association`

ewbankkit opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #10429.
Relates #12599.
Relates #15105.

Adds a new resource, `aws_vpc_endpoint_security_group_association`, similar to `aws_vpc_endpoint_route_table_association` and `aws_vpc_endpoint_subnet_association`, which allows security groups to be associated with a VPC endpoint.
The wrinkle here is that there must always be at least one security group associated with a VPC endpoint, unlike for route tables or subnets. When a VPC endpoint is created with no security groups specified the VPC's **default** security group is associated with the VPC endpoint. Prior to this PR we always required at least one security group to be specified for a VPC endpoint, so associating the **default** security group with the VPC endpoint required explicit configuration. Now associating the **default** security group with the VPC endpoint is implicit if no security groups are specified.
In order for this default association to be removed, one `aws_vpc_endpoint_security_group_association` for the VPC endpoint can be configured with `replace_default_association = true` which means that the resource will remove the default association once the specified security group is associated with the VPC endpoint and conversely, when the resource is deleted the default association is reinstated.

There are the usual caveats around not attempting to manage associations through both the `aws_vpc_endpoint` resource's `security_group_ids` attribute and individual `aws_vpc_endpoint_security_group_association` resources.

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_\|TestAccAWSVpcEndpoint_\|TestAccAWSVpcEndpointSubnetAssociation_\|TestAccAWSVpcEndpointRouteTableAssociation_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_\|TestAccAWSVpcEndpoint_\|TestAccAWSVpcEndpointSubnetAssociation_\|TestAccAWSVpcEndpointRouteTableAssociation_ -timeout 180m
=== RUN   TestAccAWSVpcEndpointRouteTableAssociation_basic
=== PAUSE TestAccAWSVpcEndpointRouteTableAssociation_basic
=== RUN   TestAccAWSVpcEndpointRouteTableAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointRouteTableAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== RUN   TestAccAWSVpcEndpointSubnetAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSubnetAssociation_basic
=== RUN   TestAccAWSVpcEndpointSubnetAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSubnetAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSubnetAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSubnetAssociation_multiple
=== RUN   TestAccAWSVpcEndpoint_gatewayBasic
=== PAUSE TestAccAWSVpcEndpoint_gatewayBasic
=== RUN   TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== RUN   TestAccAWSVpcEndpoint_gatewayPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayPolicy
=== RUN   TestAccAWSVpcEndpoint_interfaceBasic
=== PAUSE TestAccAWSVpcEndpoint_interfaceBasic
=== RUN   TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== PAUSE TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== RUN   TestAccAWSVpcEndpoint_disappears
=== PAUSE TestAccAWSVpcEndpoint_disappears
=== RUN   TestAccAWSVpcEndpoint_tags
=== PAUSE TestAccAWSVpcEndpoint_tags
=== RUN   TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
=== PAUSE TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
=== CONT  TestAccAWSVpcEndpointRouteTableAssociation_basic
=== CONT  TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpointRouteTableAssociation_basic (37.47s)
=== CONT  TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
--- PASS: TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy (69.89s)
=== CONT  TestAccAWSVpcEndpoint_tags
--- PASS: TestAccAWSVpcEndpoint_tags (85.48s)
=== CONT  TestAccAWSVpcEndpoint_disappears
--- PASS: TestAccAWSVpcEndpoint_disappears (30.97s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
--- PASS: TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer (396.86s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate (281.49s)
=== CONT  TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate (316.66s)
=== CONT  TestAccAWSVpcEndpoint_interfaceBasic
--- PASS: TestAccAWSVpcEndpoint_interfaceBasic (87.63s)
=== CONT  TestAccAWSVpcEndpoint_gatewayPolicy
--- PASS: TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup (375.61s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
--- PASS: TestAccAWSVpcEndpoint_gatewayPolicy (58.57s)
=== CONT  TestAccAWSVpcEndpoint_gatewayBasic
--- PASS: TestAccAWSVpcEndpoint_gatewayBasic (34.93s)
=== CONT  TestAccAWSVpcEndpointSubnetAssociation_multiple
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (132.97s)
=== CONT  TestAccAWSVpcEndpointSubnetAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSubnetAssociation_disappears (265.54s)
=== CONT  TestAccAWSVpcEndpointSubnetAssociation_basic
--- PASS: TestAccAWSVpcEndpointSubnetAssociation_basic (245.34s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSubnetAssociation_multiple (623.90s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (107.47s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (105.00s)
=== CONT  TestAccAWSVpcEndpointRouteTableAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (92.71s)
--- PASS: TestAccAWSVpcEndpointRouteTableAssociation_disappears (34.14s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1695.257s
```

Lightning0 commented:
This is a welcome addition to TF!

Will replace some custom code I needed to write to set up an AWS Transfer Server of type VPC which doesn't expose the VPC Endpoint and needs to use AWS Security Groups to whitelist Vendor CIDR's.

ewbankkit commented:
@Lightning0 Yes, https://github.com/terraform-providers/terraform-provider-aws/pull/12599 was the catalyst for this new resource.

ewbankkit commented:
Rebased to remove merge conflicts.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (120.46s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (122.91s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (105.48s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (114.22s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	234.769s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpoint_ -timeout 120m
=== RUN   TestAccAWSVpcEndpoint_gatewayBasic
=== PAUSE TestAccAWSVpcEndpoint_gatewayBasic
=== RUN   TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== RUN   TestAccAWSVpcEndpoint_gatewayPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayPolicy
=== RUN   TestAccAWSVpcEndpoint_interfaceBasic
=== PAUSE TestAccAWSVpcEndpoint_interfaceBasic
=== RUN   TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== PAUSE TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== RUN   TestAccAWSVpcEndpoint_disappears
=== PAUSE TestAccAWSVpcEndpoint_disappears
=== RUN   TestAccAWSVpcEndpoint_tags
=== PAUSE TestAccAWSVpcEndpoint_tags
=== CONT  TestAccAWSVpcEndpoint_gatewayBasic
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
--- PASS: TestAccAWSVpcEndpoint_gatewayBasic (35.90s)
=== CONT  TestAccAWSVpcEndpoint_tags
--- PASS: TestAccAWSVpcEndpoint_tags (65.25s)
=== CONT  TestAccAWSVpcEndpoint_disappears
--- PASS: TestAccAWSVpcEndpoint_disappears (32.93s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate (351.56s)
=== CONT  TestAccAWSVpcEndpoint_interfaceBasic
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate (316.21s)
=== CONT  TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
--- PASS: TestAccAWSVpcEndpoint_interfaceBasic (108.80s)
=== CONT  TestAccAWSVpcEndpoint_gatewayPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayPolicy (56.91s)
=== CONT  TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy (66.21s)
--- PASS: TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup (372.30s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	822.654s
```

ewbankkit commented:
Rebased and upgraded new code to Plugin SDK v2.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
    resource_aws_vpc_endpoint_security_group_association_test.go:41: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (105.14s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (107.23s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (109.48s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (111.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	111.146s
```

ewbankkit commented:
Rebased and fixed merge conflict.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (104.16s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (111.52s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
    resource_aws_vpc_endpoint_security_group_association_test.go:41: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (61.28s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (63.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	175.194s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpoint_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpoint_ -timeout 120m
=== RUN   TestAccAWSVpcEndpoint_gatewayBasic
=== PAUSE TestAccAWSVpcEndpoint_gatewayBasic
=== RUN   TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== RUN   TestAccAWSVpcEndpoint_gatewayPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayPolicy
=== RUN   TestAccAWSVpcEndpoint_interfaceBasic
=== PAUSE TestAccAWSVpcEndpoint_interfaceBasic
=== RUN   TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== PAUSE TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== RUN   TestAccAWSVpcEndpoint_disappears
=== PAUSE TestAccAWSVpcEndpoint_disappears
=== RUN   TestAccAWSVpcEndpoint_tags
=== PAUSE TestAccAWSVpcEndpoint_tags
=== CONT  TestAccAWSVpcEndpoint_gatewayBasic
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
--- PASS: TestAccAWSVpcEndpoint_gatewayBasic (39.27s)
=== CONT  TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate (272.03s)
=== CONT  TestAccAWSVpcEndpoint_tags
--- PASS: TestAccAWSVpcEndpoint_tags (94.20s)
=== CONT  TestAccAWSVpcEndpoint_disappears
--- PASS: TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup (344.51s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== CONT  TestAccAWSVpcEndpoint_disappears
    resource_aws_vpc_endpoint_test.go:445: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSVpcEndpoint_disappears (35.43s)
=== CONT  TestAccAWSVpcEndpoint_gatewayPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayPolicy (64.31s)
=== CONT  TestAccAWSVpcEndpoint_interfaceBasic
--- PASS: TestAccAWSVpcEndpoint_interfaceBasic (79.91s)
=== CONT  TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy (73.57s)
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate (326.78s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	710.605s
```

andwoods commented:
Thanks for the work on this feature! Has there been any updates or estimated time of releasing? My team could use this functionality.

kramarenko104 commented:
> Thanks for the work on this feature! Has there been any updates or estimated time of releasing? My team could use this functionality.

Our team is really looking forward to the release of this feature.  We would be glad to know the possible release time.

ewbankkit commented:
@kramarenko104 Is it attachment of security groups to an AWS Transfer Server that you require this resource for? If so, AWS recently added this functionality to the Transfer API.
I opened #15788 for this.

kramarenko104 commented:
> @kramarenko104 Is it attachment of security groups to an AWS Transfer Server that you require this resource for? If so, AWS recently added this functionality to the Transfer API.
> I opened #15788 for this.

Yes, we need it for SFTP server. Thanks, we will vote for this PR, too. 
@ewbankkit, is it difficult to say when one of these PRs can be taken into release? I'm stuck with lack of this functionality and trying now to do workaround but it looks like hard for now. So, I'm hope to see one of these PRs released ASAP.

pstrav commented:
Hey team, do we have any updates on this?

ewbankkit commented:
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_\|TestAccAWSVpcEndpoint_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_\|TestAccAWSVpcEndpoint_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== RUN   TestAccAWSVpcEndpoint_gatewayBasic
=== PAUSE TestAccAWSVpcEndpoint_gatewayBasic
=== RUN   TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== RUN   TestAccAWSVpcEndpoint_gatewayPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayPolicy
=== RUN   TestAccAWSVpcEndpoint_interfaceBasic
=== PAUSE TestAccAWSVpcEndpoint_interfaceBasic
=== RUN   TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== PAUSE TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== RUN   TestAccAWSVpcEndpoint_disappears
=== PAUSE TestAccAWSVpcEndpoint_disappears
=== RUN   TestAccAWSVpcEndpoint_tags
=== PAUSE TestAccAWSVpcEndpoint_tags
=== RUN   TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
=== PAUSE TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== CONT  TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
    resource_aws_vpc_endpoint_test.go:513: Step 1/2 error: Error running apply: 2020/11/16 08:41:22 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error waiting for VPC Endpoint (vpce-003d272b3ec2f66e7) to become available: VPC endpoint (vpce-003d272b3ec2f66e7) is in a failed state
        
        
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (98.27s)
=== CONT  TestAccAWSVpcEndpoint_tags
--- FAIL: TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer (121.50s)
=== CONT  TestAccAWSVpcEndpoint_disappears
--- PASS: TestAccAWSVpcEndpoint_disappears (31.86s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
--- PASS: TestAccAWSVpcEndpoint_tags (86.99s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate (251.34s)
=== CONT  TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate (284.20s)
=== CONT  TestAccAWSVpcEndpoint_interfaceBasic
--- PASS: TestAccAWSVpcEndpoint_interfaceBasic (106.77s)
=== CONT  TestAccAWSVpcEndpoint_gatewayPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayPolicy (59.44s)
=== CONT  TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy (69.62s)
=== CONT  TestAccAWSVpcEndpoint_gatewayBasic
--- PASS: TestAccAWSVpcEndpoint_gatewayBasic (34.98s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (90.86s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup (362.98s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (136.15s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (139.51s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	939.138s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

Same error as [here](https://github.com/hashicorp/terraform-provider-aws/pull/14012#issuecomment-727294353).

ewbankkit commented:
Rebased and resolved merge conflicts.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_\|TestAccAWSVpcEndpoint_' ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_\|TestAccAWSVpcEndpoint_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== RUN   TestAccAWSVpcEndpoint_gatewayBasic
=== PAUSE TestAccAWSVpcEndpoint_gatewayBasic
=== RUN   TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
=== RUN   TestAccAWSVpcEndpoint_gatewayPolicy
=== PAUSE TestAccAWSVpcEndpoint_gatewayPolicy
=== RUN   TestAccAWSVpcEndpoint_interfaceBasic
=== PAUSE TestAccAWSVpcEndpoint_interfaceBasic
=== RUN   TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== PAUSE TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
=== RUN   TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== PAUSE TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
=== RUN   TestAccAWSVpcEndpoint_disappears
=== PAUSE TestAccAWSVpcEndpoint_disappears
=== RUN   TestAccAWSVpcEndpoint_tags
=== PAUSE TestAccAWSVpcEndpoint_tags
=== RUN   TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
=== PAUSE TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== CONT  TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (117.84s)
=== CONT  TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer
--- PASS: TestAccAWSVpcEndpoint_interfaceWithSubnetAndSecurityGroup (384.50s)
=== CONT  TestAccAWSVpcEndpoint_tags
=== CONT  TestAccAWSVpcEndpoint_disappears
--- PASS: TestAccAWSVpcEndpoint_VpcEndpointType_GatewayLoadBalancer (335.03s)
--- PASS: TestAccAWSVpcEndpoint_tags (83.42s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate
--- PASS: TestAccAWSVpcEndpoint_disappears (30.92s)
=== CONT  TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnCreate (240.93s)
=== CONT  TestAccAWSVpcEndpoint_gatewayBasic
=== CONT  TestAccAWSVpcEndpoint_interfaceBasic
--- PASS: TestAccAWSVpcEndpoint_gatewayBasic (34.22s)
--- PASS: TestAccAWSVpcEndpoint_interfaceNonAWSServiceAcceptOnUpdate (293.72s)
=== CONT  TestAccAWSVpcEndpoint_gatewayPolicy
=== CONT  TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpoint_interfaceBasic (61.78s)
--- PASS: TestAccAWSVpcEndpoint_gatewayPolicy (60.07s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (59.76s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
--- PASS: TestAccAWSVpcEndpoint_gatewayWithRouteTableAndPolicy (66.55s)
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (52.65s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (48.86s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	936.201s
```

kchowhan commented:
any timeline on when this will be merged?

pooja1806 commented:
when can we expect this feature to be added?

pstrav commented:
any updates on this?

saiaman commented:
👍

ewbankkit commented:
##### TODO

* Will need to rebase once #18465 is merged
* Refactor `aws_vpc_endpoint_security_group_association` to use internal `finder` and `waiter` packages

DrFaust92 commented on aws/internal/service/ec2/errors.go:
```diff
@@ -55,6 +60,7 @@ const (
 )
 
 const (
+	ErrCodeInvalidSubnetIdNotFound = "InvalidSubnetId.NotFound"
```
duplicate?

ewbankkit replied:
No, this (and _Invalid route table_ and probably others) and the uppercase version are returned from different API calls 😄.

DrFaust92 replied:
😞

srspnda-instacart commented:
Hey @ewbankkit we've been tracking this branch for awhile now, using it successfully in production. Any idea how long before we get this merged into upstream?

ewbankkit commented:
```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSVpcEndpointSecurityGroupAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSVpcEndpointSecurityGroupAssociation_ -timeout 180m
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== PAUSE TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_basic
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_multiple
=== CONT  TestAccAWSVpcEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_ReplaceDefaultAssociation (86.03s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_basic (90.93s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_disappears (93.30s)
--- PASS: TestAccAWSVpcEndpointSecurityGroupAssociation_multiple (94.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	98.082s
```

Ninir commented:
@DrFaust92 Would it be possible to get this one reviewed 🙏 It's about to cross the finish line. Thanks!

Shocktrooper commented:
Wee just learned that our organization prevents us from doing the workaround of directly attaching security groups to Network Interfaces so an actual attachment the right way with a terraform resource would go a long way to solve our issues

ewbankkit commented:
#### Commercial

```console
% make testacc TESTARGS='-run=TestAccEC2VPCEndpointSecurityGroupAssociation_\|TestAccEC2VPCEndpoint_interfaceBasic' PKG=ec2 ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 3  -run=TestAccEC2VPCEndpointSecurityGroupAssociation_\|TestAccEC2VPCEndpoint_interfaceBasic -timeout 180m
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_basic
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation
=== RUN   TestAccEC2VPCEndpoint_interfaceBasic
=== PAUSE TestAccEC2VPCEndpoint_interfaceBasic
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_basic
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_basic (107.05s)
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation (112.27s)
=== CONT  TestAccEC2VPCEndpoint_interfaceBasic
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_multiple (117.03s)
--- PASS: TestAccEC2VPCEndpoint_interfaceBasic (76.91s)
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_disappears (83.11s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	193.958s
```

#### GovCloud

```console
% make testacc TESTARGS='-run=TestAccEC2VPCEndpointSecurityGroupAssociation_\|TestAccEC2VPCEndpoint_interfaceBasic' PKG=ec2 ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 3  -run=TestAccEC2VPCEndpointSecurityGroupAssociation_\|TestAccEC2VPCEndpoint_interfaceBasic -timeout 180m
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_basic
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_basic
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_disappears
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_disappears
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_multiple
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_multiple
=== RUN   TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation
=== PAUSE TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation
=== RUN   TestAccEC2VPCEndpoint_interfaceBasic
=== PAUSE TestAccEC2VPCEndpoint_interfaceBasic
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_basic
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation
=== CONT  TestAccEC2VPCEndpoint_interfaceBasic
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_replaceDefaultAssociation (99.86s)
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_multiple
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_basic (105.70s)
=== CONT  TestAccEC2VPCEndpointSecurityGroupAssociation_disappears
--- PASS: TestAccEC2VPCEndpoint_interfaceBasic (105.92s)
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_multiple (90.74s)
--- PASS: TestAccEC2VPCEndpointSecurityGroupAssociation_disappears (92.38s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	201.647s
```

YakDriver commented on website/docs/r/vpc_endpoint_security_group_association.html.markdown:
```diff
@@ -0,0 +1,43 @@
+
+* `security_group_id` - (Required) The ID of the security group to be associated with the VPC endpoint.
+* `vpc_endpoint_id` - (Required) The ID of the VPC endpoint with which the security group will be associated.
+* `replace_default_association` - (Optional) Whether this association should replace the association with the VPC's default security group that is created when no security groups are specified during VPC endpoint creation.
+At most 1 association per-VPC endpoint should be configured with `replace_default_association = true`.
```
```suggestion
* `replace_default_association` - (Optional) Whether this association should replace the association with the VPC's default security group that is created when no security groups are specified during VPC endpoint creation. At most 1 association per-VPC endpoint should be configured with `replace_default_association = true`.
```
Tiny nit: Less likely to lose a piece if it's all on one line. So later someone doesn't come in and add a new argument in between and then you've got an orphan.
